fix: right click in clickable statusline

famiu opened the pull request:
Previously, clickable statuslines didn't receive right click unless the statusline was focused. This fixes that behavior by making clickable statusline sections correctly receive right clicks regardless of whether the statusline is focused or not.

Closes #18994

famiu commented:
Wondering if `status_line_offset` and `sep_line_offset` are good names for the variables that previously used to be `on_status_line` and `on_sep_line`, since those variables don't necessarily mean that the mouse is actually on the status line (or sep line) but rather indicate an offset from it.

justinmk commented on src/nvim/mouse.c:
```diff
@@ -189,58 +191,72 @@ int jump_to_mouse(int flags, bool *inclusive, int which_button)
+  if (keep_focus) {
+    // If we can't change focus, set the value of row and col back to absolute values
+    // since the values relative to the window are only used when keep_focus is true
+    row = prev_row;
```
the comment is confusing because keep_focus is true here.

famiu replied:
the comment was incorrect, I fixed it now

justinmk commented on src/nvim/mouse.c:
```diff
@@ -128,8 +128,9 @@ bool is_mouse_key(int c)
 {
-  static int on_status_line = 0;        // #lines below bottom of window
-  static int on_sep_line = 0;           // on separator right of window
+  static int status_line_offset = 0;        // vertical offset from status line
```
```suggestion
  static int status_line_offset = 0;        // #lines offset from status line
```

justinmk commented on src/nvim/mouse.c:
```diff
@@ -128,8 +128,9 @@ bool is_mouse_key(int c)
-  static int on_status_line = 0;        // #lines below bottom of window
-  static int on_sep_line = 0;           // on separator right of window
+  static int status_line_offset = 0;        // vertical offset from status line
+  static int sep_line_offset = 0;           // horizontal offset from sep line
```
```suggestion
  static int sep_line_offset = 0;           // #cols offset from sep line
```

justinmk commented on src/nvim/mouse.c:
```diff
@@ -189,58 +191,70 @@ int jump_to_mouse(int flags, bool *inclusive, int which_button)
-    }
+  if (row < 0 || col < 0) {                   // check if it makes sense
+    return IN_UNKNOWN;
+  }
```
the docstring explains the `if (!(flags & MOUSE_FOCUS))` guard as follows, is this still the behavior?

```
/// If flags has MOUSE_FOCUS, then the current window will not be changed, and
/// if the mouse is outside the window then the text will scroll, or if the
/// mouse was previously on a status line, then the status line may be dragged.
```

famiu replied:
yes, unless the statusline or winbar was clicked, then there will be no scroll

justinmk commented on src/nvim/mouse.c:
```diff
@@ -189,58 +191,70 @@ int jump_to_mouse(int flags, bool *inclusive, int which_button)
+  } else {
+    on_status_line = false;
+  }
+  if (row == -1) {
+    on_winbar = wp->w_winbar_height != 0;
+  } else {
+    on_winbar = false;
+  }
```
```suggestion
  on_winbar = (row == -1)
    ? wp->w_winbar_height != 0
    : false;
```

justinmk commented:
https://github.com/neovim/neovim/issues/18994#issuecomment-1176961485 gives a perfect, succinct explanation of the issue: 

> * Right-click won't work unless you left-click first
> * The modifier does not work, nothing happens when I do shift right click

Does shift right click work with this PR? Need a test for that.

famiu committed:
fix: make right click work correctly with clickable statusline

Previously, clickable statuslines didn't receive right click unless the statusline was focused. This fixes that behavior by making clickable statusline sections correctly receive right clicks regardless of whether the statusline is focused or not.

Closes #18994
